crates/fulf/src/interface/mod.rs

```diff
@@ -0,0 +1,637 @@
+    /// can slow down the task significantly. And there's very few reasons
+    /// for a line to exceed, for example, 1024 bytes:
+    ///
+    /// 1. This is a line in a text that is not code.
+    ///
+    /// 2. This is a non-formatted line of automatically generated code.
+    ///
+    /// 3. This is a very bad code.
+    ///
+    /// 4. Some very rare other reasons, like giant right-shifted branching.
+    ///
+    /// And in any of those cases there's probably no point in fuzzing such line.
+    #[inline]
+    pub fn with_fzy_algo(
+        path: impl AsRef<Path>,
+
+        needle: impl Into<Box<str>>,
+        max_line_len: usize,
+
+        sort_and_print: impl FnMut(&mut [MWP], usize),
+    ) -> Option<(Vec<MWP>, usize)> {
+        let needle = needle.into();
+
+        if needle.is_empty() || needle.len() > max_line_len {
+            return Default::default();
+        }
+
+        let r = {
+            let mut r = Rules::new();
+            r.bonus_threads = if cfg!(target_pointer_width = "64") {
+                2
+            } else {
+                0
+            };
+
+            r
+        };
+
+        let path = path.as_ref();
+        let dir_iter = ignore::Walk::new(path);
+
+        let root_folder = path.to_str()?;
+
+        let is_ascii = needle.is_ascii();
+
+        Some(if is_ascii {
+            // ascii
+            let fzydata = FzyData::new(root_folder.into(), needle, max_line_len);
+            FzyAscii::setter(dir_iter, fzydata, r, sort_and_print)
+        } else {
+            // utf8
+            let algo = move |needle: &str, line: &str| {
+                if line.len() > max_line_len {
+                    None
+                } else {
+                    utf8::match_and_score_with_positions(needle, line)
+                }
+            };
+
+            let data = GenericSearchData::new(root_folder.into(), needle, algo);
+            GenericAlgo::setter(dir_iter, data, r, sort_and_print)
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{showcase::*, *};
+    use std::time::{Duration, SystemTime};
+
+    #[test]
+    fn basic_functionality_test() {
+        const DELAY: Duration = Duration::from_secs(2);
+        let mut past = SystemTime::now();
+
+        let sort_and_print = |results: &mut [MWP], total| {
+            results.sort_unstable_by(|a, b| b.1.cmp(&a.1));
+
+            let now = SystemTime::now();
+
+            if let Ok(dur) = now.duration_since(past) {
+                if dur > DELAY {
+                    past = now;
+
+                    for idx in 0..8 {
+                        if let Some(pack) = results.get(idx) {
+                            let (s, _score, pos) = pack;
+                            println!("Total: {}\n{}\n{:?}", total, s, pos);
+                        } else {
+                            break;
+                        }
+                    }
+                }
+            }
+        };
```

Comment:
This part in the test shows the signature of the required `sort_and_print` function.

But due to the multithreading of the whole thing, it's different from the one-threaded dynamic filter; there won't be as many prints as there is in one-threaded filter due to the overhead of synchronization of many threads. I will improve the synchronization routine soon, but the main problem will still be here: you can only get results when the filter produces `results_cap` items. Probably, this filter should beat the rarity of updates with better general speed.

Reply:
We could make `results_cap` be 128, or even 64, that's totally fine.

Reply:
> We could make `results_cap` be 128, or even 64, that's totally fine.

Yes... but at the same time, no. Less results to store - blocking thread synchronization is done more often. Actually, 128 or 64 should work almost as good as 512, but I'm not sure about the real cost.

Reply:
Haven't really used it, so I'm also not sure if 512 or 128 has a different effect noticeably.